test/test_query_geo.cpp

```diff
@@ -149,7 +149,7 @@ TEST(Query_GeoWithinBasics)
     auto&& location = table->column<Link>(location_column_key);
 
     CHECK_EQUAL(location.geo_within(GeoBox{GeoPoint{0.2, 0.2}, GeoPoint{0.7, 0.7}}).count(), 1);
-    CHECK_EQUAL(location.geo_within(GeoBox{GeoPoint{-2, -1.5}, GeoPoint{0.7, 0.5}}).count(), 3);
+    CHECK_EQUAL(location.geo_within(GeoBox{GeoPoint{-2, -1.5}, GeoPoint{0.7, 0.5}}).count(), 2);
```

Comment:
Weird... These results change quite a bit to a few very weird ones.
Do you have any idea why is with this: 
```
diff --git a/test/test_query_geo.cpp b/test/test_query_geo.cpp
index 6a2b3c410..32cb8ece0 100644
--- a/test/test_query_geo.cpp
+++ b/test/test_query_geo.cpp
@@ -29,6 +29,7 @@
 #include <realm/group.hpp>
 #include <realm/table.hpp>
 #include <realm/query_expression.hpp>
+#include <realm/table_view.hpp>

 #include <ostream>
 #include <sstream>
@@ -165,6 +166,21 @@ TEST(Query_GeoWithinBasics)
     CHECK_THROW_CONTAINING_MESSAGE(location.geo_within(Geospatial{}),
                                    "Invalid region in GEOWITHIN query for parameter 'NULL': 'NULL cannot be used on "
                                    "the right hand side of a GEOWITHIN query");
+
+    auto dump_query = [&] (Query &&q) {
+        auto tv = q.find_all();
+        std::cout << "-- " << q.get_description_safe() << " -> " << tv.size() << " keys";
+        for (size_t i = 0; i < tv.size(); ++i)
+            std::cout << "\n\t" << i << ": " << tv[i].to_string();
+        std::cout << std::endl;
+    };
+    dump_query(location.geo_within(GeoBox{GeoPoint{0, 0}, GeoPoint{0.5, 0.5}}));
+    dump_query(location.geo_within(GeoBox{GeoPoint{0, 0}, GeoPoint{0.5, 1}}));
+    dump_query(location.geo_within(GeoBox{GeoPoint{0, -0.5}, GeoPoint{0.5, 1}}));
+    dump_query(location.geo_within(GeoBox{GeoPoint{-2, -1.5}, GeoPoint{0.7, 0.5}}));
+    dump_query(location.geo_within(GeoBox{GeoPoint{-2, -1}, GeoPoint{1, 0.5}}));
+    dump_query(location.geo_within(GeoBox{GeoPoint{-2, -2}, GeoPoint{1, 1}}));
+    dump_query(location.geo_within(GeoBox{GeoPoint{-2, -2}, GeoPoint{0.5, 1}}));
 }

 TEST(Geospatial_ListOfPrimitives)
```

original results through S2LatLng are:
```
-- location GEOWITHIN GeoBox([0, 0], [0.5, 0.5]) -> 2 keys
        0: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        1: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoBox([0, 0], [0.5, 1]) -> 2 keys
        0: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        1: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoBox([0, -0.5], [0.5, 1]) -> 2 keys
        0: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        1: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoBox([-2, -1.5], [0.7, 0.5]) -> 3 keys
        0: {"_key":0,"_id":0,"location":{"_key":0,"type":"Point","coordinates":[-2.0000000000000000e+00,-1.0000000000000000e+00]}}
        1: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        2: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoBox([-2, -1], [1, 0.5]) -> 3 keys
        0: {"_key":0,"_id":0,"location":{"_key":0,"type":"Point","coordinates":[-2.0000000000000000e+00,-1.0000000000000000e+00]}}
        1: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        2: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoBox([-2, -2], [1, 1]) -> 4 keys
        0: {"_key":0,"_id":0,"location":{"_key":0,"type":"Point","coordinates":[-2.0000000000000000e+00,-1.0000000000000000e+00]}}
        1: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        2: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
        3: {"_key":4,"_id":4,"location":{"_key":4,"type":"Point","coordinates":[1.0000000000000000e+00,1.0000000000000000e+00]}}
-- location GEOWITHIN GeoBox([-2, -2], [0.5, 1]) -> 3 keys
        0: {"_key":0,"_id":0,"location":{"_key":0,"type":"Point","coordinates":[-2.0000000000000000e+00,-1.0000000000000000e+00]}}
        1: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        2: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
```

... and with the patch they are quite different:
```
-- location GEOWITHIN GeoPolygon({[0, 0], [0, 0.5], [0.5, 0.5], [0.5, 0], [0, 0]}) -> 0 keys
-- location GEOWITHIN GeoPolygon({[0, 0], [0, 1], [0.5, 1], [0.5, 0], [0, 0]}) -> 1 keys
        0: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoPolygon({[0, -0.5], [0, 1], [0.5, 1], [0.5, -0.5], [0, -0.5]}) -> 0 keys
-- location GEOWITHIN GeoPolygon({[-2, -1.5], [-2, 0.5], [0.7, 0.5], [0.7, -1.5], [-2, -1.5]}) -> 2 keys
        0: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        1: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoPolygon({[-2, -1], [-2, 0.5], [1, 0.5], [1, -1], [-2, -1]}) -> 2 keys
        0: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        1: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoPolygon({[-2, -2], [-2, 1], [1, 1], [1, -2], [-2, -2]}) -> 4 keys
        0: {"_key":0,"_id":0,"location":{"_key":0,"type":"Point","coordinates":[-2.0000000000000000e+00,-1.0000000000000000e+00]}}
        1: {"_key":1,"_id":1,"location":{"_key":1,"type":"Point","coordinates":[-1.0000000000000000e+00,-2.0000000000000000e+00]}}
        2: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        3: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
-- location GEOWITHIN GeoPolygon({[-2, -2], [-2, 1], [0.5, 1], [0.5, -2], [-2, -2]}) -> 4 keys
        0: {"_key":0,"_id":0,"location":{"_key":0,"type":"Point","coordinates":[-2.0000000000000000e+00,-1.0000000000000000e+00]}}
        1: {"_key":1,"_id":1,"location":{"_key":1,"type":"Point","coordinates":[-1.0000000000000000e+00,-2.0000000000000000e+00]}}
        2: {"_key":2,"_id":2,"location":{"_key":2,"type":"Point","coordinates":[0.0000000000000000e+00,0.0000000000000000e+00]}}
        3: {"_key":3,"_id":3,"location":{"_key":3,"type":"Point","coordinates":[5.0000000000000000e-01,5.0000000000000000e-01]}}
```

Shouldn't S2LatLng be equivalent to what you wrote for transformation to S2Polygon?

Reply:
The difference appears to be how a S2LatLngRect implements it's ::Contains() method. It is assumed that in the rectangle points along the edge vertices are included, while that is not necessarily the case in a Polygon.